#DE Analyses Function--------
# Author: Brian J. Sanderson <brian.sanderson@ttu.edu>
#
# Adapted from a script written by Jack Colicchio <jmcbulls89@gmail.com>.
# Colicchio JM, Monnahan PJ, Kelly JK, Hileman LC. 2015. Gene expression
# plasticity resulting from parental leaf damage in Mimulus guttatus.
# New Phytologist 205: 894â€“906.
#
#
# <counttable> is a dataframe of count data, such as that generated by
# htseq-count
#
# <group1> is the number of samples in the first comparison group (e.g. males)
# <group2> is the number of samples in the second comparison group
# (e.g. females)
#
# Returns a dataframe of the significant results from limma, edgeR, and
# DeSeq2 at a B-H FDR < 0.01. Also reports the number of genes that
# have > two-fold FC values at that same significance threshold.
#

runDGE <- function(counttable, group1, group2) {
  require("DESeq2")
  require("limma")
  require("edgeR")
  #---------DeSeq
  # Convenience variables
  countmat<-as.matrix(counttable)
  row.names(countmat)<-row.names(counttable)

  # Make the design matrix
  counttableDE<-as.matrix(counttable)
  colData <- data.frame(condition=c(rep("C", times=group1),
				    rep( "D", times=group2)))

  #Prepare data object, estimate normalization factors
  suppressMessages(cds <-DESeqDataSetFromMatrix(countData = counttableDE,
						colData=colData,
						design=formula(~condition)))
  cds <- estimateSizeFactors( cds )
  cds <- estimateDispersions( cds )

  #Run DeSeq, extract and format results
  suppressMessages(cds <- DESeq(cds))
  suppressMessages(res <- results(cds))
  res<-as.matrix(res)
  res<-data.frame(res)

  #identifying genes with an FDR<0.01
  resSig <- res[ res$padj < 0.01 & !is.na(res$padj), ]
  res<-as.matrix(res)
  AllData<-merge(res, countmat, by=0)

  #---------Limma

  # Calculate norm factors
  nf <- calcNormFactors(counttable, method="TMM")

  #Voom conversion of data
  meow<-voom(counttable)

  #Running limma
  wow<-lmFit(meow, design=cbind(Grp1=1,Grp2vs1=c(rep(0, times=group1),
						 rep( 1, times=group2))))

  #bayseian determination of liklihood of DE
  wow<-eBayes(wow)
  yow<-topTable(wow,coef=2, number=nrow(counttable))

  #finding genes with FDR<0.01 by limma
  resSiglimma <- yow[ yow$adj.P.Val < 0.01 & !is.na(yow$adj.P.Val), ]
  m<-nrow(resSiglimma)
  AllData<-merge(yow, AllData, by.x=0, by.y=1, all=T)

  #---------Edge R

  #Prepare data object, calculate norm factors
  y<-DGEList(counttable)
  y <-calcNormFactors(y, method="TMM")

  #Design matrix
  expR<- factor(c(rep(0, times=group1), rep( 1, times=group2)))
  design <- model.matrix(~expR)

  y <- estimateGLMTrendedDisp(y, design)

  #using most effective dispersion estimate, depend on trended
  #disp estimate preceeding it.
  y <- estimateGLMTagwiseDisp(y, design)
  fit<-glmFit(y, design)
  lrt<-glmLRT(fit, coef=2)

  #finding genes with FDR<0.01
  FDR <- p.adjust(lrt$table$PValue, method="BH")
  FDR<-as.matrix(FDR)
  logFC_ER <- lrt$table$logFC
  logFC_ER <- as.matrix(logFC_ER)
  row.names(FDR)<-row.names(counttable)
  row.names(logFC_ER) <- row.names(counttable)
  FDR<-data.frame(FDR)
  resSigEdger <- FDR[ FDR$FDR < 0.01 & !is.na(FDR$FDR), ]#FDR<0.01
  resSigEdger<-data.frame(resSigEdger)
  ES<- nrow(resSigEdger)

  #merging files
  AllData<-merge(FDR, AllData, by.x=0, by.y=1, all=T)
  AllData<-merge(logFC_ER, AllData, by.x=0, by.y=1, all=T)

  summed <- nrow(AllData[AllData$FDR<0.01 &
                           AllData$adj.P.Val<0.01 &
                           AllData$padj < 0.01 &
                           !is.na(AllData$padj),])

  summedThres <- nrow(AllData[AllData$FDR<0.01 &
                                abs(AllData$logFC) > 1 &
                                AllData$adj.P.Val<0.01 &
                                abs(AllData$log2FoldChange) > 1 &
                                AllData$padj < 0.01 &
                                !is.na(AllData$padj) &
                                abs(AllData$V1) > 1,])

  cat(paste("You have", summed, "genes differentially expressed across all analyses",    "\n"))
  cat(paste("You have", summedThres, "genes differentially expressed across all analyses, with a 2-fold change",    "\n"))

  #Return the consensus set of differentially expressed genes
  return(AllData[AllData$FDR<0.01 &
                   AllData$adj.P.Val<0.01 &
                   AllData$padj < 0.01 &
                   !is.na(AllData$padj),])
}
