# DE Analyses Function--------
# Author: Brian J. Sanderson <brian.sanderson@ttu.edu>
#
# Adapted from a script written by Jack Colicchio <jmcbulls89@gmail.com>.
# Colicchio JM, Monnahan PJ, Kelly JK, Hileman LC. 2015. Gene expression
# plasticity resulting from parental leaf damage in Mimulus guttatus.
# New Phytologist 205: 894-906.
#
# Estimate differential gene expression between two groups using DESeq2, limma, and edgeR.
# <counttable> is a dataframe of count data, such as that generated by
# htseq-count
# 
# <counttable>: A dataframe of RNA-seq read count data, such as that generated htseq-count. 
#               The row should be gene names.
#
# <group1>: The number of samples in the first comparison group (e.g. males). 
#
# <group2>: The number of samples in the second comparison group (e.g. females).
# 
# <FDRthresh>: The Benjamini-Hochberg false discovery rate value to use as a significance threhold 
#              for calling genes differentially expressed (default is 0.01). 
#              Only used if fullResults=FALSE.
# 
# <fullResults>: A parameter that specifies whether to return a dataset that includes the results 
#                for all genes regardless of significance (TRUE) or only the results that meet the 
#                significance threshold across all three analyses (FALSE). 
#                Default value is FALSE.
#
# Returns the log2FC and B-H FDR values from DESeq2, limma, and edgeR, either for all 
# genes (fullResults=TRUE) or only genes that meet the significance threshold (fullResults=FALSE).


runDGE <- function(counttable, group1, group2, FDRthresh=0.01, fullResults=FALSE) {
  require("DESeq2")
  require("limma")
  require("edgeR")
  #---------DESeq2
  # Convenience variables
  countmat<-as.matrix(counttable)
  row.names(countmat)<-row.names(counttable)

  # Make the design matrix
  counttableDE<-as.matrix(counttable)
  colData <- data.frame(condition=c(rep("C", times=group1),
                                    rep( "D", times=group2)))

  #Prepare data object, estimate normalization factors
  suppressMessages(cds <-DESeqDataSetFromMatrix(countData = counttableDE,
                                                colData=colData,
                                                design=formula(~condition)))
  suppressMessages(cds <- estimateSizeFactors( cds ))
  suppressMessages(cds <- estimateDispersions( cds ))

  #Run DeSeq, extract and format results
  suppressMessages(cds <- DESeq(cds))
  suppressMessages(res <- results(cds))
  res<-as.matrix(res)
  
  #Merge DESeq2 results with the count matrix
  AllData<-merge(res, countmat, by=0)

  #---------Limma

  # Calculate norm factors
  nf <- calcNormFactors(counttable, method="TMM")

  #Voom conversion of data
  meow<-voom(counttable)

  #Running limma
  wow<-lmFit(meow, design=cbind(Grp1=1,Grp2vs1=c(rep(0, times=group1),
                                                 rep( 1, times=group2))))

  #Bayseian determination of liklihood of DE
  wow<-eBayes(wow)
  yow<-topTable(wow,coef=2, number=nrow(counttable))

  #Merge limma results with the results from DESeq2
  AllData<-merge(yow, AllData, by.x=0, by.y=1, all=T)

  #---------edgeR

  #Prepare data object, calculate norm factors
  y<-DGEList(counttable)
  y <-calcNormFactors(y, method="TMM")

  #Design matrix
  expR<- factor(c(rep(0, times=group1), rep( 1, times=group2)))
  design <- model.matrix(~expR)

  y <- estimateGLMTrendedDisp(y, design)

  #Using most effective dispersion estimate, depend on trended
  #disp estimate preceeding it.
  y <- estimateGLMTagwiseDisp(y, design)
  fit<-glmFit(y, design)
  lrt<-glmLRT(fit, coef=2)

  #Calculate FDR and logFC for edgeR analysis
  FDR <- p.adjust(lrt$table$PValue, method="BH")
  FDR<-as.matrix(FDR)
  logFC_ER <- lrt$table$logFC
  logFC_ER <- as.matrix(logFC_ER)
  row.names(FDR)<-row.names(counttable)
  row.names(logFC_ER) <- row.names(counttable)
  FDR<-data.frame(FDR)
  
  #Merge edgeR results with the results from DESeq2 and limma
  AllData<-merge(FDR, AllData, by.x=0, by.y=1, all=T)
  AllData<-merge(logFC_ER, AllData, by.x=0, by.y=1, all=T)
  
  #Give columns better names
  colnames(AllData)[c(1,2,3,4,8,11,15)] <- c("gene", "edgeR_logFC", "edgeR_FDR", "limma_logFC", "limma_FDR", "DESeq2_logFC", "DESeq2_FDR")
  AllData[,1] <- as.character(AllData[,1])
  
  
  
  #A convenience variable to plot genes that are significant across all analyses
  
  AllData$sig <- FALSE
  AllData$sig[AllData$edgeR_FDR<FDRthresh &
                       AllData$limma_FDR<FDRthresh &
                       AllData$DESeq2_FDR<FDRthresh &
                       !is.na(AllData$DESeq2_FDR)] <- TRUE
  #Log CPM for plotting
  AllData$logCPM <- mglmOneGroup(counttableDE)
  
  #Plot logFC vs logCPM
  maPlot(x=NULL, y=NULL, logAbundance=AllData$logCPM, logFC=AllData$edgeR_logFC, de.tags=which(AllData$sig),
         smooth.scatter=FALSE, lowess=FALSE, pch=19, cex=0.5, smearWidth=0.5, panel.first=grid(),
         ylab=bquote(paste("edgeR ", log[2],"FC")), xlab=bquote(paste(log[2], CPM)))

  #Determine how many genes pass the significance threshold in all 3 analyses
  summed <- length(which(AllData$sig))

  #Determine how many genes pass the significance threshold and a log2FC > 1 cut off
  summedThres <- nrow(AllData[AllData$sig == TRUE &
                                abs(AllData$edgeR_logFC) > 1 &
                                abs(AllData$limma_logFC) > 1 &
                                abs(AllData$DESeq2_logFC) > 1,])
  
  cat(paste("You have", summed, "genes differentially expressed across all analyses",    "\n"))
  cat(paste("You have", summedThres, "genes differentially expressed across all analyses, with a 2-fold change",    "\n"))
  
  outData <- AllData[,c(1,2,3,4,8,11,15)]

  if(fullResults==TRUE) { #Return just the gene names, logFC, and B-H FDR values for all genes regardless of significance
    return(outData)
  } else { #Return the gene names, logFC, and B-H FDR values for genes that meet signifciance threshold across all analyses
    outData <- outData[outData$edgeR_FDR < FDRthresh &
                         outData$limma_FDR < FDRthresh &
                         outData$DESeq2_FDR < FDRthresh &
                         !is.na(outData$DESeq2_FDR),]
    return(outData)
  }
  
}
